package/kernel/linux/modules/other.mk

```diff
@@ -869,6 +869,85 @@ endef
 $(eval $(call KernelPackage,zram))
 
 
+define KernelPackage/zsmalloc
+  SUBMENU:=$(OTHER_MENU)
+  TITLE:=ZSMALLOC support
+  DEPENDS:=+kmod-crypto-deflate \
+	+kmod-lib-lz4 \
+	@!PACKAGE_kmod-zram
+  KCONFIG:= \
+	CONFIG_ZSMALLOC \
+	CONFIG_ZSMALLOC_STAT=n
+  FILES:= $(LINUX_DIR)/mm/zsmalloc.ko
+  AUTOLOAD:=$(call AutoLoad,19,zsmalloc)
+endef
+
+define KernelPackage/zsmalloc/description
+ Special purpose memory allocator for compressed memory pages
+endef
+
+define KernelPackage/zsmalloc/config
+	if PACKAGE_kmod-zsmalloc
+		config KERNEL_PGTABLE_MAPPING
+			bool "zsmalloc: enable CONFIG_PGTABLE_MAPPING"
+			default y if arm
+			default n
+			help
+	  Enable CONFIG_PGTABLE_MAPPING in the kernel for faster memory
+	  allocations when using ZSMALLOC, in some architectures. Enabled
+	  by default for the ARM architecture because it may be a huge
+	  performance boost.
+	endif
+endef
+
+$(eval $(call KernelPackage,zsmalloc))
+
+
+define KernelPackage/zram-writeback
+  SUBMENU:=$(OTHER_MENU)
+  TITLE:=zram with writeback support
+  DEPENDS:=+kmod-zsmalloc
+  KCONFIG:= \
+	CONFIG_ZRAM \
+	CONFIG_ZRAM_DEBUG=n \
+	CONFIG_ZRAM_MEMORY_TRACKING=n \
+	CONFIG_ZRAM_WRITEBACK=y
+  FILES:= \
+	$(LINUX_DIR)/drivers/block/zram/zram.ko
+  AUTOLOAD:=$(call AutoLoad,20,zram)
+endef
+
+define KernelPackage/zram-writeback/description
+ Compressed RAM disk with support for page writeback
+endef
+
+$(eval $(call KernelPackage,zram-writeback))
+
+
+define KernelPackage/zswap
+  SUBMENU:=$(OTHER_MENU)
+  TITLE:=zswap compressed swapping cache
+  DEPENDS:=+kmod-zsmalloc
+  KCONFIG:= \
+	CONFIG_FRONTSWAP=y \
+	CONFIG_Z3FOLD \
+	CONFIG_ZBUD \
+	CONFIG_ZPOOL \
+	CONFIG_ZSWAP=y
```

Comment:
zswap is not a module and therefore there is no solution about that problem. You don't have to believe me, explore the kernel source code. The same is true for frontswap. The memory allocators and the zpool are indeed modules. If you think about a solution (removing zswap is not a solution, because it doesn't solve the problem, just to clarify), I'll like to know it.

Reply:
An interesting thing about the build bot is that this package is in conflict with `zram-swap`, and it will never be built automatically and your comment, that the kernel will always build the `zswap` code is, therefore, false.

The code is designed to *eventually* totally eradicate the `zram-swap` package because it is so bad in quality that I really doubt how it is still in the official repositories.

And I'm not judging for free the package, I did created a way better solution ready to end the life of `zram-swap`.

Reply:
> zswap is not a module and therefore there is no solution about that problem.

This features increases resulting 5.4.28 zImage by 7344 bytes on imx6/cortex-a9 so I guess, that it's going to be similar on other targets.

I'll put aside, that this is very niche feature usable only on a few targets/devices, but I think, that it's not acceptable to include this config options by default on targets constrained by the SMALL_FLASH constraint etc. IIRC most of those targets don't use swap at all, because they don't have suitable block device for use etc.

> zpool are indeed modules

zpool is also built-in once the zswap is built-in, means always.

Reply:
Is this the only change needed?
Because if we talk about the default config, the packages will never be built by default as long as `zram-swap` exist. But I can trivially add the negative dependency, not a big deal. The packages are idependent (not as @CodeFetch somehow thinks) and without `zswap`, it will only use the advanced features for `zram`.

As a side note: please note that the new `zram` enables `lz4hc` and `zstd`, so for SMALL_FLASH they need somehow space to exist.

Let me know what else is needed for this to be merged.

Reply:
> I'll put aside, that this is very niche feature usable only on a few targets/devices, but I think, that it's not acceptable to include this config options by default on targets constrained by the SMALL_FLASH constraint etc. IIRC most of those targets don't use swap at all, because they don't have suitable block device for use etc.

I have something to add here, @ynezz:
The idea behind using `zswap` is to be a cache for `zswap`. You don't need a backing storage and that's why they share a configuration: they are meant to be used torgerther.
Why? Let me explain:
1. `zswap` is a cache, while `zram` is a compressed RAM disk.
2. Because of how the `zpool` works compared to how the compressed block device work, `zswap` *is faster*.
3. The idea is to use the faster `zswap` in front (that's why the kernel feature is called FRONTSWAP) is to save time compared to the time required to do the same work in `zram`, which is a direct winning in performance.
4. In the back of `zswap` is `zram` with a slower but does a way better job at compressing, so we have a direct win on resource usage.
5. The two layers will provide the best of them, one being faster and the other using a better algorithm for compression.
6. That is true even if no backing storage is used. But the benefits are event greater if it exists because great amounts of IO will be saved and, when the memory cannot be compressed, great amounts of incomprensible unused data will be moved from memory to the backing storage, freeing even more memory.

So `zswap` is not meant to be used as a niche feature, but as direct a performance benefit. If small flash devices cannot run with the extra modules, I think it's better to disable this package completely instead of only disabling `zswap`.